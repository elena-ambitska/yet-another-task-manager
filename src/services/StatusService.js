import BaseService from "./BaseService";

class StatusService extends BaseService {
    async getStatuses() {
        return await this.get(`statuses`);
    }
}

export default new StatusService();
