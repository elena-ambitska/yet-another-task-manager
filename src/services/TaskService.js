import cards_list from "../tests/mock/cards_list";

class TaskService {
    list() {
        return new Promise((resolve) => {
            resolve(cards_list);
        })
    }
}

export default new TaskService()