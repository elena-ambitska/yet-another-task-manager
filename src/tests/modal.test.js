import React from "react";
import Modal from "../components/modal";
import TaskService from "../services/TaskService";
import updateCard from "./mock/updateCard";
import StatusService from "../services/StatusService";
import statuses from "./mock/statuses";
import {renderWithContext} from "./render";

import {act, fireEvent, screen} from "@testing-library/react";
import createCard from "./mock/createCard";
import createCard_fail_400 from "./mock/createCard_fail_400";


const createCardMock = jest.spyOn(TaskService, "createCard");
const updateCardMock = jest.spyOn(TaskService, "updateCard");
const listStatusMock = jest.spyOn(StatusService, "getStatuses");

beforeEach(() => {
    createCardMock.mockClear();
    updateCardMock.mockClear();
    listStatusMock.mockClear();

    listStatusMock.mockReturnValueOnce(Promise.resolve(statuses));
});

const modalMock = (result) => {
    return async (dispatch) => {
        return Promise.resolve(result);
    };
};

const setActiveMock = (active) => {

};

describe("Test Modal component", () => {
    it("should render modal component", () => {
        renderWithContext(<Modal currentTask={{"id": 1541}}/>);
        expect(screen.findByRole(/document/i)).not.toBeNull();
    });

    it("modal component sends data to services", async () => {
        createCardMock.mockReturnValueOnce(modalMock(createCard));

        renderWithContext(<Modal active={true} setActive={setActiveMock} currentTask={{}}/>);

        await act(() => {
            fireEvent.click(screen.getByText(/Save changes/i));
        })

        expect(createCardMock).toHaveBeenCalledTimes(1);
    });

    it("modal component displays error when fails", async () => {
        createCardMock.mockReturnValueOnce(modalMock(createCard_fail_400));

        renderWithContext(<Modal active={true} setActive={setActiveMock} currentTask={{}}/>);
        await act(() => {
            fireEvent.click(screen.getByText(/Save changes/i));
        })

        expect(createCardMock).toHaveBeenCalledTimes(1);

        const errorList = await screen.getByRole("errors-list");
        expect(errorList.children.length).toBe(createCard_fail_400.data.errors.title.length);
        expect(errorList.children[0]).toHaveTextContent("title must be defined.");
    });

    it("modal component put and update data to services", async () => {
        updateCardMock.mockReturnValueOnce(modalMock(updateCard));

        renderWithContext(<Modal currentTask={
            {
                id: 12,
                title: "React",
                description: "Hello"
            }} active={true} setActive={setActiveMock}
        />);
        await act(() => {
            fireEvent.click(screen.getByText(/Save changes/i));
        })

        expect(updateCardMock).toHaveBeenCalledTimes(1);
    });


    it("modal component closed on button clicked", async () => {
        const setActiveMock = jest.fn();

        renderWithContext(<Modal active={true} setActive={setActiveMock} currentTask={{}}/>);

        await act(() => {
            fireEvent.click(screen.getByRole("close"));
        })
        await act(() => {
            fireEvent.click(screen.getByRole("btn-secondary"));
        })

        await act(() => {
            fireEvent.click(screen.getByRole("dialog"));
        })

        expect(setActiveMock).toHaveBeenCalledWith(false);
        expect(setActiveMock).toHaveBeenCalledTimes(3);
    });
});
