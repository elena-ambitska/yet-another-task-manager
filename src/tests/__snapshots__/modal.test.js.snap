// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Modal component should render modal component 1`] = `
<div
  className="modal fade"
  onClick={[Function]}
  role="dialog"
  style={
    Object {
      "display": "none",
    }
  }
  tabIndex="-1"
>
  <div
    className="modal-dialog"
    onClick={[Function]}
    role="document"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h5
          className="modal-title"
        >
          Create card
        </h5>
        <button
          aria-label="Close"
          className="close"
          data-dismiss="modal"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </button>
      </div>
      <div
        className="modal-body"
      >
        <form>
          <div
            className="mb-3"
          >
            <label
              className="form-label"
              htmlFor="titleInput"
            >
              Title
            </label>
            <input
              className="form-control"
              id="titleInput"
              name="title"
              onChange={[Function]}
              required={true}
              value=""
            />
          </div>
          <div
            className="mb-3"
          >
            <label
              className="form-label"
              htmlFor="descriptionInput"
            >
              Description
            </label>
            <input
              className="form-control"
              id="descriptionInput"
              name="description"
              onChange={[Function]}
              required={true}
              value=""
            />
          </div>
          <label
            htmlFor="exampleFormControlSelect1"
          >
            Status
          </label>
          <select
            className="form-control"
            id="exampleFormControlSelect1"
            name="status"
            onChange={[Function]}
            value="to_do"
          >
            <option
              key="to_do"
              value="to_do"
            >
              to_do
            </option>
            <option
              key="in_progress"
              value="in_progress"
            >
              in_progress
            </option>
            <option
              key="testing"
              value="testing"
            >
              testing
            </option>
            <option
              key="done"
              value="done"
            >
              done
            </option>
          </select>
          <div
            className="modal-footer"
          >
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="submit"
            >
              Save changes
            </button>
            <button
              className="btn btn-secondary"
              data-dismiss="modal"
              onClick={[Function]}
              type="button"
            >
              Close
            </button>
            <ul
              className="errors-list"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
